packages/server/prod-server/src/libs/render/utils.ts

```diff
@@ -33,3 +38,46 @@ export const injectServerDataStream = (
     }),
   );
 };
+
+export const runServerRender = async ({
+  ctx,
+  serverContext,
+  bundleJS,
+  enableVM,
+}: {
+  ctx: ModernServerContext;
+  bundleJS: string;
+  enableVM?: boolean;
+  serverContext: SSRServerContext;
+}) => {
+  if (enableVM) {
+    const global = {
+      __SERVER_RENDER__: async (_: SSRServerContext) => '',
+      require,
+      console,
+      Buffer,
+      // eslint-disable-next-line node/no-unsupported-features/node-builtins, node/prefer-global/url-search-params
+      URLSearchParams,
+      // eslint-disable-next-line node/no-unsupported-features/node-builtins, node/prefer-global/url
+      URL,
+      AbortController,
+      Headers,
+      Request,
+      Response,
+      global: {},
+      exports: {},
+      module: {},
+    };
+    runInNewContext(
+      `
+${await fs.readFile(bundleJS, 'utf-8')}
+globalThis.__SERVER_RENDER__ = module.exports.serverRender;
+    `,
+      global,
+    );
+    return global.__SERVER_RENDER__(serverContext);
```

Comment:
Should this content need to cached with `cache` function?

Reply:
And I think `global.__SERVER_RENDER__` should use `runInNewContext` for finally running `renderToString` in new global.